lib/msf/base/sessions/winrm_command_shell.rb

```diff
@@ -0,0 +1,174 @@
+# -*- coding: binary -*-
+
+require 'winrm'
+
+module Msf::Sessions
+  #
+  # This class provides a session for WinRM client connections, where Metasploit
+  # has authenticated to a remote WinRM instance.
+  #
+  class WinrmCommandShell < Msf::Sessions::PowerShell
+
+    def commands
+      {
+        'help'       => 'Help menu',
+        'background' => 'Backgrounds the current shell session',
+        'sessions'   => 'Quickly switch to another session',
+        'resource'   => 'Run a meta commands script stored in a local file',
+        'irb'        => 'Open an interactive Ruby shell on the current session',
+        'pry'        => 'Open the Pry debugger on the current session',
+      }
+    end
+
+    #
+    # Create an MSF command shell from a WinRM shell object
+    #
+    # @param shell [WinRM::Shells::Base] A WinRM shell object
+    # @param opts [Hash] Optional parameters to pass to the session object.
+    def initialize(shell, opts = {})
+      self.shell = shell
+      # To buffer input received while a session is backgrounded, we stick responses in a list
+      @buffer_mutex = Mutex.new
+      @buffer = []
+      @check_stdin_event = Rex::Sync::Event.new(false, true)
+      super(nil, opts)
+    end
+
+    def shell_write(buf)
+      return unless buf
+
+      begin
+        framework.events.on_session_command(self, buf.strip)
+        shell.send_stdin(buf)
+      rescue ::Rex::SocketError, ::EOFError, ::IOError, ::Errno::EPIPE => e
+        shell_close
+        raise e
+      end
+    end
+
+  ##
+  # :category: Msf::Session::Provider::SingleCommandShell implementors
+  #
+  # Read from the command shell.
+  #
+  def shell_read(length=-1, timeout=1)
+    result = ""
+    @buffer_mutex.synchronize {
+      result = @buffer.join("")
+      @buffer = []
+      if length > -1 and result.length > length
+        # Return up to length, and keep the rest in the buffer
+        extra = result[length..-1]
+        result = result[0,length]
+        @buffer << extra
+      end
+    }
+    result
+  end
```

Comment:
```suggestion
```

I think we could delete these method overrides if we implement this approach: https://github.com/rapid7/metasploit-framework/pull/15632/files#r705357961